[many] TLS

danking opened the pull request:
This PR teaches most of our cluster how to exclusively speak HTTPS instead of
HTTP. The exceptions are:
- from batch-driver to batch workers
- from batch workers to internal-gateway
- to ukbb-rg
- from router to notebook workers
- letsencrypt (oh the irony)

## Changes from Original PR Proposal

### Root Certificate

I added a secret to default named `ssl-config-hail-root` containing `hail-root-key.pem`, and `hail-root-cert.pem`. Every principal trusts this root. This root trusts every principal. This PR originally prevented clients from speaking to servers with certs they didn't trust. Now everyone trusts everyone. As long as the root key is not leaked this is OK. Only `create_certs` mounts this secret. The key is used to sign every certificate and the cert is included in each principal's incoming and outgoing trust lists. The root certificate and key are never re-created, so our deploys have no downtime and we avoid addressing the rotation problem.

I removed all the trust specifications. A later PR will resolve rotation and mTLS. That PR will restore the trust specifications. I didn't change the structure of the secrets (they still have an incoming and outgoing trust list which only contains the root cert) because I need this structure for mTLS anyway.


The original PR text follows.

---

## HTTPS and TLS

HTTP is implemented on TCP/IP. HTTPS is also implemented on TCP/IP and differs
very mildly. After the socket is opened, a TLS [1] connection is established
over the socket. Thereafter, every HTTP message is encrypted and transported by
the TLS machinery. The HTTP protocol is unchanged. The default port fort HTTP is
80 and the default port for HTTPS 443, however any port may be used.

There are currently four versions of TLS, the latest is TLS 1.3. All versions of
SSL are considered insecure.

The OpenSSL library implements TLS. There are other implementations, such as
LibreSSL, but they implement roughly the same interface as OpenSSL.

The TLS protocol defines both a scheme for the *encryption of messages* and for
the *authentication of parties*. The protocol defines authentication as
optional. In practice, at least one party presents authentication. For example,
public web servers authenticate themselves to clients but clients do not
reciprocate.

### Authentication
#### X.509 Certificates

TLS uses X.509 Certificates for authentication. X.509 is a rather complicated
standard. X.509 Certificates can be serialized in a variety of ways. We use the
Privacy-enhanced Electronic Mail (PEM) file format for serialization. PEM is
really simple. A file may contain multiple base64-encoded blobs each with a
header and footer of the form:

```
-----BEGIN LABEL-----
...
-----END LABEL-----
```

where `LABEL` describes the data. We only use two labels: `CERTIFICATE` and
`PRIVATE KEY`.

An X.509 Certificate is an unforgeable proof of identity. It usually is paired
with a private key that was used to digitally sign the certificate. In the
security literature, an authenticatable entity is usually called a
*principal*. Each principal should have a unique private key. In our system the
principals are both our services (e.g. `batch`, `batch-driver`) and any
non-serving clients (e.g. `test-batch`, `admin-pod`). A key and certificate are
generated ad nihilum by `openssl req -new`:

```
openssl req -new \
        -x509
        -keyout key_file
        -out cert_file
        -newkey rsa:4096
        -nodes
        -subj /CN=example.com
        -addext subjectAltName = DNS:www.example.com,DNS:foo.com
```

The first three arguments are self-explanatory. I explain the rest:

- `-newkey rsa:4096`. TLS supports many different kinds of private keys. This
  generates a 4096-bit private key.
- `-nodes`. This should be read as "no DES". It means that the private key is not
  itself encrypted using DES and a password.
- `-subj /CN=example.com`. This certificate is valid for a server whose DNS name
  is `example.com`. If "hostname checking" is enabled (web browsers always
  enable it), then the client will reject the certificate if the hostname used
  to open the socket does not match the certificate.
- `-addext subjectAltName = ...`. This specifies additional acceptable hostnames
  for the aformentioned hostname check.

#### Trust

An X.509 Certificate only proves that the principal has a private key. On the
world wide web, ownership of a particular domain, e.g. `example.com`, is
guaranteed by a Certificate Authority (CA). A Certificate Authority (like
letsencypt or VeriSign) digitally signs the certificate of a user that has
proven they own the domain name mentioned in the certificate (see above
discussion of `CN` and `subjectAltName`). Web browsers have a file of the "root"
certificates of the public Certificate Authorities. They establish a "chain of
trust" from a root certificate to the server's certificate.

Our system is simpler. We have no root certificate. Each principal has a
certificate which is given to all the principals to which it might
communicate.

### Encryption

TLS has many encryption schemes. I will focus on encryption using a *symmetric*
key because asymmetric schemes do not enable *forward secrecy* [2]. Under
forward secret schemes, the two parties share a private key unknown to all
adversaries. TLS 1.3 makes forward secrecy mandatory. I intend to eventually
require all our services to refuse to speak anything other than TLS 1.3.

The shared private key is used to encrypt and decrypt messages sent over a
socket. This poses a problem: how do two parties who have never met each other
agree on a private key without revealing the key to the public? This is a
classic cryptography problem called [key
exchange](https://en.wikipedia.org/wiki/Key_exchange). The classic solution to
this problem is [Diffie-Hellman key
exchange](https://en.wikipedia.org/wiki/Diffie–Hellman_key_exchange). The
Wikipedia article has "General overview" which is quite clear.

In addition to a key, the parties must agree on a cipher. There are many old,
insecure ciphers available. In the future I intend all our servers to refuse to
use insecure ciphers. Mozilla
[has a list of secure cipher suites](https://wiki.mozilla.org/Security/Server_Side_TLS#Recommended_configurations).

## New Hail Concepts

Every principal in our system has a secret: `ssl-config-NAME`. These secrets are
automatically created for a particular namespace by `tls/create_certs.py`. Who
trusts who (i.e. who is allowed to talk to whom) is defined by
`tls/config.yaml`. For example, `site` is defined in `config.yaml` as follows:

```
- name: site
  domain: site
  kind: nginx
  incoming:
  - admin-pod
  - router
```

A principal named "site" exists. Site's domain names are `site`,
`site.NAMESPACE`, `site.NAMESPACE.svc.cluster.local`. Site's configuration files
should be in NGINX configuration file format. Site accepts incoming requests
from the principals named admin-pod and router. Site is not permitted to make
any outgoing requests. `create_certs.py` will create a new secret named
`ssl-config-site` which contains five files:

- `site-config-http.conf`: an NGINX configuration file that configures TLS for
  incoming requests.
- `site-config-proxy.conf`: an NGINX configuration file that configures TLS for
  outgoing (proxy_pass) requests.
- `site-key.pem`: a private key.
- `site-cert.pem`: a certificate.
- `site-incoming.pem`: a list of certificates trusted for incoming requests.
- `site-outgoing.pem`: a list of certificates expected from outgoing requests.

If site makes an HTTP request to a server and that server does not return a
certificate in `site-outgoing.pem`, it will immediately halt the connection. I
intend (though do not currently) site to also reject incoming requests that are
not accompanied by a certificate in `site-incoming.pem`. I describe the [trouble
with that later](#incoming-trust).

There are two other kinds: `json` and `curl`. The former is for Hail Python
services. The later is for the admin-pod and image-fetcher.

Deploy will run `create_certs` on every master deploy. Newly deployed services
will be unable to talk to not-yet-deployed services. I include the
one-deploy-ago certificates in the trust chains, but once incoming trust is
fixed, I am unsure how to smoothly upgrade services. I probably need to notify
old services to refresh their certificates after the secrets are updated.

### Incoming Trust

Mutual TLS (mTLS) refers to TLS connections wherein both sides are
authenticated. This is rare on the web. In our system, it means verifying that a
request made to you carries a certificate in the `NAME-incoming.pem` file. I
cannot enable that in this PR because the three unmanaged services,
router-resolver, internal-gateway, and gateway, do not currently have
certificates. As a result, all the services in the PR namespace reject the
requests from the unmangaed services. In particular, batch pods cannot
communicate with batch-driver.

After this PR is deployed and the unmanaged services have certificates, I can
enable mutual TLS. I've marked the tow lines that need to change with `# FIXME:
mTLS`.

### Batch Confused Deputy

The [confused deputy
problem](https://en.wikipedia.org/wiki/Confused_deputy_problem) is a classic
in computer security. It refers to a situation with two parties: the deputy and
the attacker. The deputy has authority that the attacker does not. For example,
the deputy can arrest people. A *confused deputy* is one which has been tricked
by the attacker into misusing its authority.

In master, Batch has a confused deputy problem: it issues a callback in response
to a batch finishing. That callback is issued from within the cluster and
therefore can name many of our services which are not exposed to the
internet. With the introduction of TLS everywhere, a confused deputy callback
will fail because the victim will not receive a valid Batch certificate (batch
purposely does not send certificates with the callbacks). Batch only uses its
certificate to send a callback for CI. This is safe because we control CI and
ensure it is not compromised.

In the long run, I want to fix batch to use an entirely different network for
callbacks.

### Notes of Annoyance

`aiohttp` silently ignores most invalid TLS requests, this makes debugging a TLS
issue difficult.

`aiohttp`'s `raise_for_status` does not include the HTTP body in the
message. NGINX sometimes returns 400 in response to TLS issues with a proxy. It
includes crucial details in the HTTP body. I usually debug these issues by
sshing to the client pod and using curl to manually reproduce the error.

Readiness and liveness probes cannot use HTTP. Although k8s supports HTTPS, it
does not support so-called "mTLS" or "mutual TLS." This is fancy verbiage for
servers that require clients to send trusted certificates. I will eventually require
this. There is a lot of information in GitHub issues and the Istio web pages
about this, but at the end of the day, kubernetes does not support this. TCP
probes are the best we can do. There [was a
PR](kubernetes/kubernetes#61231) to allow httpGet probes
to send the kubelet certificate, but it was closed because, apparently, the
[httpGet probes can be targeted at arbitrary IP
addresses](kubernetes/kubernetes#61231 (review)) (what
the hell?), ergo Confused Deputy.


### Future Work

- Require TLS 1.3 everywhere.
- Comply with Mozilla's "Modern" recommendations.
- [Incoming Trust](#incoming-trust).
- Refresh certificates after deploying new ones.

### Footnotes

[1] TLS: Transport Layer Security. Preceded by Secure Sockets Layer (SSL) which
    is not obsolete and insecure. After SSL version 3, a new version of SSL was
    proposed in RFC 2246. This new version was backwards-incompatible and was
    thus given a new name: Transport Layer Security. In common discussion, SSL
    and TLS are used interchangeable. Indeed, the python TLS library is called
    `ssl`.

[2] Forward secrecy is a property of an encryption system. Forward secrecy means
    a message cannot be decrypted in the future by an adversary who learned one
    of the private keys. For example, imagine you are sending sensitive messages
    to another individual. If that individual is later coerced into revealing
    their secret key, forward secrecy would prevent the coercer from reading
    your messages. Forward secrecy is achieved by negotiating a shared private
    key between the two parties that is only used for a "session" and then
    discarded. If the session key is securely discarded and neither key can
    recreate it without cooperation from the other key, then *one* leaked key is
    insufficient to reveal the messages.

---

Things for you to verify:
- does every service load *its own unqiue* ssl-config secret?
- does every service use an SSL context for serving?
- does everyone who makes http requests to internal services use an SSL client
  session?
- do all TLS-secured nginx instances include their http.conf in the `http`
  section and the `proxy.conf` file in any proxying sections?
- Do the SSLContext's from `ssl.py` and the nginx configurations generated by
  `create_certs.py` actually ensure security?

Post-merge actions:
- deploy gateway
- deploy internal-gateway
- deploy router-resolver

Anticipated outages:

- Before a service is redeployed it will be inaccessible from the outside
  because the router will try to speak to it on HTTPS. Services that speak to
  one another (ci<->batch, everyone<->auth) will lose connections while the
  deploy is happening. Deploy should move smoothly because CI will completely
  transmit the deploy batch to batch before batch goes dark.
- dev namespaces will be broken until the owner redeploys the router, etc.

cseed commented:
> There are other implementations, such as LibreSSL, but they implement roughly the same interface as OpenSSL.

LibreSSL introduced a new interface, libtls, that's designed to be easy to use, secure by default, and match the underlying socket semantics as much as possible.  If we ever do any C level networking, we should use it.

 > ad nihilum

You mean ex nihilo, from nothing?

 > I intend to eventually require all our services to refuse to speak anything other than TLS 1.3.

Can we get a task for this in Asana?  And for mTLS?

 > Our system is simpler. We have no root certificate.
 > Deploy will run create_certs on every master deploy.
 > [O]nce incoming trust is fixed, I am unsure how to smoothly upgrade services.

I think we should a have a root certificate for all services in the cluster and verify all certificates are signed by the root certificate.  I don't know how to handle creating new certs on every deploy, either.  I think we should just create them if they don't already exist.  It looks like you're pinning keys for incoming and outgoing, which is awesome.  It looks like you're duplicating the keys for each incoming/outgoing list it appears in.  Alternatively, you could break the cert secret into two parts: the private key/config needed by the service, and the certs needed by the clients/servers.

 > In the long run, I want to fix batch to use an entirely different network for callbacks.

I agree.  Can we get a task for this?  Using authorization to control who can talk to whom is great.  We should enforce that with network policies.  Defense in depth.  Another task.

 > Readiness and liveness probes cannot use HTTP.
 > Although k8s supports HTTPS, it does not support so-called "mTLS" or "mutual TLS."

Ugh.  But probes can be run via a command, so we should be able to use curl and our client certs for this: https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-startup-probes/

```
    livenessProbe:
      exec:
        command:
        - cat
        - /tmp/healthy
      initialDelaySeconds: 5
      periodSeconds: 5
```

cseed commented:
Also, thanks for that detailed write up.  That was incredibly clear and instructive.

danking commented:
Ah, yes, ex nihilio, should've taken a latin class.

I added these to uncharted:
- [security] enable mTLS for all services
- [security] disable TLS <1.3
- [security] comply with Mozilla's "modern" recommendations
- [batch][security] use a separate network for batch's callbacks


### liveness probes

Ah, that's a good point. I'll rewrite to use curl and the client's own certificate and I'll make sure clients trust themselves.

### root cert

I don't think it is possible in aiohttp to both verify a certificate has a valid chain from a root cert and, separately, exists in a list of trusted certificates. The effect would be that every client would trust every server because every server certificate is signed by the same root certificate. I think using a root cert is quite secure (a big improvement over our current situation!). However, I endeavored in this PR to additionally prevent, for example, a compromised `notebook` from masquerading as `batch`.

I agree that additionally verifying that the certificate came from a single root certificate (that we, perhaps, destroy after everything is signed) would additionally prevent a malicious user from inserting their certificates into the trusted certificates list. AFAICT, python's `ssl` module has no support for this verification strategy. We could probably build an SSLContext shim that contained two SSLContexts one with a root cert and one with the trusted certs and require certification verification to pass both. Seems easy to get wrong, so I'm inclined to not take this path.

### trusted cert lists

Yeah, it felt a little silly to duplicate the cert in each secret. However, this seems like the simplest approach if I require each principal to only trust a subset of incoming/outgoing principals. If I had one secret per principal, then I have to modify build.yaml or deployment.yamls if I modify the trust sets. That seemed error prone. If I had one secret with all the certs, then when a service starts up it has to select the trusted ones and only insert those into its certificate store. This seems OK, but a little harder to inspect.

Duplicating a cert for each trust list to which it belongs occupies what seems like a good spot to me from a developer ergonomics perspective:
- O(trusts) modifications necessary to update/revoke the cert
- O(1) configuration to load a trust list
- no pod-start-time configuration
- the trust list is on the container's file system, so its easy to inspect

Small point: I don't pin the incoming certs yet due to the mTLS challenges.

### create on each deploy

Only creating certs if they don't exist is an easy change. Seems fine, though leaves unresolved how to rotate the certs. I guess I'm inclined to always recreate because it makes rotation the common case, forcing us to make it work well.

I think the only way to do a no-downtime rotation is:
1. create fresh certs
2. create the trust lists including a principal's fresh cert and previous generation cert
3. update all the secrets
4. somehow ensure everyone has the latest secrets?
5. notify all servers to refresh their certificates (nginx: send SIGHUP, aiohttp: we have to write something)

We could stick a generation uuid in the secrets and keep refreshing services until the certificate uuid they read is the one our deploy expects.

### mTLS

This PR will land. Things will break because the unmanaged services (router-resolver, gateway, internal-gateway) do not speak TLS. I'll manually deploy them. The default namespace and new PR namespaces should now function properly. Developers will need to redeploy from master.

With this in place, I will make another PR with two main changes:
- enable client verification, and
- modify create_certs.py to load the unmanaged certificates from `default` rather than the local namespace.
That PR should pass all the tests (batch pods will speak TLS to internal-gateway; internal-gateway will speak TLS to PR batch using a client certificate PR batch trusts; etc.). Merge that PR. Everything will function correctly; however, the unmanaged services will not verify client certificates. I manually deploy and now everyone is verifying client certificates.

danking commented:
@cseed 
I added a secret to default named `ssl-config-hail-root` containing `hail-root-key.pem`, and `hail-root-cert.pem`. Every principal trusts this root. This root trusts every principal. This PR originally prevented clients from speaking to servers with certs they didn't trust. Now everyone trusts everyone. As long as the root key is not leaked this is OK. Only `create_certs` mounts this secret. The key is used to sign every certificate and the cert is included in each principal's incoming and outgoing trust lists. The root certificate and key are never re-created, so our deploys have no downtime and we avoid addressing the rotation problem.

I removed all the trust specifications. A later PR will resolve rotation and mTLS. That PR will restore the trust specifications. I didn't change the structure of the secrets (they still have an incoming and outgoing trust list which only contains the root cert) because I need this structure for mTLS anyway.


I've updated the PR description with this text so it ends up in the squashed commit.

danking commented:
be-rump

cseed commented on hail/python/hailtop/ssl.py:
```diff
@@ -0,0 +1,229 @@
+
+    def request(self, *args, **kwargs):
+        if 'ssl' not in kwargs:
+            kwargs['ssl'] = self.ssl_context
```
I'm quite surprised this is necessary.  Can't you set this when creating the session?

cseed commented on hail/python/hailtop/ssl.py:
```diff
@@ -0,0 +1,229 @@
+    log.info(f'using tls and verifying client and server certificates')
+
+
+class TLSAdapter(HTTPAdapter):
```
I feel like you can do this by assigning the verify and cert attributes of requests.Session, no?

cseed commented on tls/create_certs.py:
```diff
@@ -0,0 +1,193 @@
+    http_config_file = f'ssl-config-http.conf'
+    proxy_config_file = f'ssl-config-proxy.conf'
+    with open(proxy_config_file, 'w') as proxy, open(http_config_file, 'w') as http:
+        proxy.write(f'proxy_ssl_certificate         /ssl-config/{cert};\n')
```
A multi-line string is probably more readable.

cseed commented on tls/config.yaml:
```diff
@@ -0,0 +1,52 @@
+  kind: nginx
+- name: router-resolver
+  domain: router-resolver
+  unmanged: True
```
Unmanaged means it is running of http?

cseed commented on tls/create_certs.py:
```diff
@@ -0,0 +1,193 @@
+    domain = p['domain']
+    unmanaged = p.get('unmanaged', False)
+    if unmanaged and namespace != 'default':
+        return
```
Don't quite understand why default is special here.

cseed commented on tls/create_certs.py:
```diff
@@ -0,0 +1,193 @@
+        sp.check_call(['kubectl', 'apply', '-f', k8s_secret.name])
+
+
+def download_previous_certs():
```
What's this doing?
